Use IntDict, use ints in more places, and clean up recip code.

showell opened the pull request.

andersk commented on static/js/fold_dict.ts:
```diff
@@ -0,0 +1,79 @@
+
+/**
+ * Implementation detail of the FoldDict class. `key` is `k` converted to a string,
+ * in lowercase if the `fold_case` option is enabled.
```
With `fold_case` no longer an option, the comment needs a small update.

andersk commented on static/js/fold_dict.ts:
```diff
@@ -0,0 +1,79 @@
+};
+
+/**
+ * This class primarily exists to support the fold_case option, because so many
```
And here.

andersk commented on static/js/people.js:
```diff
@@ -837,24 +838,25 @@ exports.get_rest_of_realm = function get_rest_of_realm() {
-    return false;
+    const ids = duplicate_full_name_data.get(full_name);
+
+    return ids && ids.size > 1;
```
You mean `length`?

showell replied:
it's a set :)

andersk commented on static/js/people.js:
```diff
@@ -837,24 +838,25 @@ exports.get_rest_of_realm = function get_rest_of_realm() {
-        ids.set(user_id);
+        ids.add(user_id);
     }
     if (to_remove && user_id && ids.has(user_id)) {
```
Could skip this `.has` call.

andersk commented on frontend_tests/zjsunit/index.js:
```diff
@@ -16,6 +16,7 @@ global._ = require('underscore/underscore.js');
 const _ = global._;
 
 global.Dict = require('../../static/js/dict').Dict;
+global.FoldDict = require('../../static/js/fold_dict').FoldDict;
```
I’d prefer to keep `FoldDict` out of `global` and import it just in the files where it’s used. (`Dict` is special because it’s used in `zjsquery.js`—although really I’d like to get it out of `global` too along with everything else, it’s just a bit more work.)

andersk commented on .eslintrc.json:
```diff
@@ -14,6 +14,7 @@
         "Dict": false,
         "FetchStatus": false,
         "Filter": false,
+        "FoldDict": false,
```
(Assuming we keep `FoldDict` out of `global`, remove this too.)

timabbott commented on frontend_tests/node_tests/dict.js:
```diff
@@ -64,7 +64,9 @@ run_test('undefined_keys', () => {
-    d = new Dict({fold_case: true});
+    blueslip.set_test_data('error', 'Tried to call a FoldDict method with an undefined key.');
+
+    d = new FoldDict();
```
Do we want to retain a test class for `Dict` after this migration?  Or is the goal to remove `Dict` and replace it with built-in things?

showell replied:
We really haven't changed the level of coverage here, since all we're effectively doing is making `FoldDict` a synonym for `Dict({fold_case: true}`, and we retained 100% line coverage.  But we may have never been truly covering both branches for all possible values.

I think we're fine on coverage for now (we get lots of drive-by coverage from other tests).

But to your second question, I think we want to replace plain Dict first with builtins, and when we do that, we probably want to scrutinize the tests a bit harder.

andersk replied:
`dict.ts` is currently marked as `enforce_fully_covered` in `tools/test-js-with-node`, but the fact that you didn’t get an error about forgetting to add `fold_dict.ts` means that it’s missing some coverage. Specifically, `.items`, `.num_items`, and `.clear` are uncovered. I understand that it’s “effectively” covered by virtue of being copied from a file that was covered, but we’re about to delete that file.

(Also, maybe these days `enforce_fully_covered` should be the default and we should have a list of non-fully-covered files instead, so we don’t add new non-fully-covered files without realizing it?)

showell replied:
Ok, I'm adding `fold_dict.ts` now to the list of fully-covered files and splitting up the tests.

I have no major objection to making the list be a blacklist instead of a whitelist.  I think the blacklist would be still kinda long, but I like the notion that you have to opt out for new files (especially since new files tend to be kinda small and not hard to test-cover).

showell commented:
@andersk Thanks for the review!  I'll try to address your findings tomorrow.  One quick thing is that `Set` uses `size`, not `length`, which kinda makes sense if you think about it:

~~~ js
> x = new Set([1,2,3])
Set { 1, 2, 3 }
> x.size
3
> x.length
undefined

showell commented:
@andersk I think I've addressed all your comments.  Can you give it another review?

showell commented:
I've added some commits here.  The `IntDict` commit is pretty invasive, because I am greedy about wanting maps to be keyed by ints and not have to convert them to strings.

I think Anders will be happy with the zjsunit-related commits. :)

andersk commented on static/js/message_list_view.js:
```diff
@@ -552,6 +552,7 @@ MessageListView.prototype = {
                 // mention text to show the user's current name,
                 // assuming that you're not searching for text
                 // inside the highlight.
+                user_id = parseInt(user_id, 10);
```
Perhaps this conversion should be moved into `get_user_id_for_mention_button` (which would then return `"*" | number`).

showell replied:
yeah, that's fair, since in the `email` path you already have an int...changed

andersk commented on static/js/people.js:
```diff
@@ -117,9 +124,7 @@ exports.is_known_user_id = function (user_id) {
-    user_ids = _.map(user_ids, function (user_id) {
-        return parseInt(user_id, 10);
-    });
+    user_ids = map_int(user_ids);
```
This conversion probably belongs in the caller.

showell replied:
The story there was there were lots of callers, all passing in tiny list of user_ids.  I fixed it anyway.  In a couple cases, the caller already had ints, so no need to convert.  In the other cases, it wasn't completely clear that they were ints, so I convert them.

andersk commented on static/js/people.js:
```diff
@@ -155,7 +160,9 @@ exports.huddle_string = function (message) {
 exports.user_ids_string_to_emails_string = function (user_ids_string) {
-    const user_ids = user_ids_string.split(',');
+    let user_ids = user_ids_string.split(',');
+    user_ids = map_int(user_ids);
```
Prefer `const user_ids = map_int(user_ids_string.split(','))`—the type-shifting mutation is unnecessary, and TypeScript won’t like it when it becomes TypeScript.

showell replied:
fixed

andersk commented on static/js/people.js:
```diff
@@ -241,6 +248,7 @@ exports.get_user_time = function (user_id) {
 };
 
 exports.get_user_type = function (user_id) {
+    user_id = parseInt(user_id, 10);
```
This conversion probably belongs in the caller.

andersk commented on static/js/people.js:
```diff
@@ -297,7 +305,8 @@ exports.get_recipients = function (user_ids_string) {
     const user_ids = user_ids_string.split(',');
-    const other_ids = _.reject(user_ids, exports.is_my_user_id);
+    let other_ids = _.reject(user_ids, exports.is_my_user_id);
+    other_ids = map_int(other_ids);
```
This conversion should be moved up to `const user_ids = map_int(user_ids_string.split(','))`.

andersk commented on static/js/presence.js:
```diff
@@ -38,7 +38,11 @@ exports.get_status = function (user_id) {
+    user_ids = _.map(user_ids, function (s) {
+        return parseInt(s, 10);
+    });
+
```
Prefer `const user_ids = Object.keys(exports.presence_info).map(s => parseInt(s, 10))`. The type-shifting mutation is unnecessary, and TypeScript won’t like it when it becomes TypeScript.

Perhaps `presence_info` should be `IntDict` to avoid this conversion entirely?

showell replied:
Oops, I missed this one.  I'm not sure presence is the best candidate for `IntDict`, since it's a short-lived data structure coming from the server as raw JSON every minute or so.  Gonna fix the `let`.

andersk commented on static/js/stream_create.js:
```diff
@@ -364,7 +364,8 @@ exports.create_handlers_for_users = function (container) {
                 const elem = $(this);
-                const user_id = elem.attr('data-user-id');
+                let user_id = elem.attr('data-user-id');
+                user_id = parseInt(user_id, 10);
```
Prefer `const user_id = parseInt(elem.attr('data-user-id'), 10)`. The type-shifting mutation is unnecessary, and TypeScript won’t like it when it becomes TypeScript.

showell replied:
fixed

andersk commented on static/js/activity.js:
```diff
@@ -127,8 +127,14 @@ exports.get_huddles = function () {
+    return _.map(huddle.split(','), function (s) {
+        return parseInt(s, 10);
+    });
+}
```
When written as `huddle.split(',').map(s => parseInt(s, 10))`, this doesn’t seem worth extracting as a helper function.

showell replied:
Keeping this as is.  Makes it easy to look at the two callers and see what code is in common.

showell replied:
I actually changed it as a separate commit (not pushed yet).

andersk commented on static/js/people.js:
```diff
@@ -17,24 +18,30 @@ exports.init = function () {
+    return _.map(lst, function (s) {
+        return parseInt(s, 10);
+    });
+}
```
When written as `lst.map(s => parseInt(s, 10))`, this doesn’t seem worth extracting as a helper function.

showell replied:
keeping as is, `parseInt` has evil behavior if you call it wrong, I want to encapsulate it

andersk commented on static/js/fold_dict.ts:
```diff
@@ -0,0 +1,85 @@
+    // Handle case-folding of keys and the empty string.
+    private _munge(key: string): string | undefined {
+        if (key === undefined) {
+            blueslip.error("Tried to call a FoldDict method with an undefined key.");
```
I assume we’re reasonably confident this doesn’t happen, so this should use `throw new TypeError` instead of `blueslip.error`. Exceptions are reliably caught by blueslip now, after #12737.

showell replied:
I think most of the time this has happened in the past, it's a pretty recoverable error, usually due to some race condition.  And if it never happens, what's the difference?  Can we defer this?

andersk replied:
I want to be able to trust our types, instead of being forever worried about what “should” happen if we’re trying to continue execution after something has already gone wrong due to an unknown type bug. This trust is only possible when bugs are consistently squashed, not accommodated.

@timabbott What do you think?

showell replied:
(Addressed in comment in main conversation.)

andersk commented on static/js/int_dict.ts:
```diff
@@ -0,0 +1,76 @@
+    }
+
+    clear(): void {
+        this._map = new Map();
```
`this._map.clear()` seems more direct.

showell replied:
fixed

showell commented on static/js/int_dict.ts:
```diff
@@ -0,0 +1,76 @@
+    convert it to a Set, not an IntDict.
+*/
+
+export class IntDict<integer, V> {
```
ah, of course, I fixed this, thanks

showell commented on static/js/int_dict.ts:
```diff
@@ -0,0 +1,76 @@
+
+        if (typeof key !== 'number') {
+            blueslip.error("Tried to call a IntDict method with a non-integer.");
+        }
```
I am gonna restructure this code to be a bit more defensive and convert "5" to 5 if it happens (but still report an error, so we can fix the upstream code as appropriate).

showell commented:
Addressed all the feedback and re-pushed.  There are a couple places where I disagree on inlining helper functions.  We can always do that as a follow-up.  I fixed all the unnecessary two-liners that caused a `let` as well as other things pointed out, like moving some conversions earlier in the code flow.

I also made the code more defensive.  So "5" will work as a key; we'll just whine about it.

showell commented:
Also added a few more commits.

showell commented:
> 
> 
> Apparently the primary source of stringified `stream_id`s is `data-stream-id` attributes. We can use `elem.data("stream-id")` instead of `parseInt(elem.attr("data-stream-id"), 10)`, since [`.data`](https://api.jquery.com/data/) already converts to `number` where appropriate.

<rant>I really wish we didn't put so much data in the DOM, but that's how the settings code work.</rant>

I'll fix it to use `data('stream-id')` (and double-check that it does convert to int).

showell commented on static/js/people.js:
```diff
@@ -385,6 +394,7 @@ exports.all_user_ids_in_pm = function (message) {
         return elem.user_id || elem.id;
     });
 
+    user_ids = map_int(user_ids);
```
Ok, that's fair.  I was tempted to kill off the `sort_numerically` there, but JS sorts [42, 5] lexically, even though the numbers are elements.  We don't need the sort to actually be numerical, but we need it to be consistent in all places (so you map the same huddle to the same bucket every time).

Then the other thing is that we then join the userids into a string.

I'm sort of inclined to cross the TS bridge when we get to it, but I'll think about this.

showell replied:
Ok, I killed off `map_int`.  In two places I now call `split_to_ints`.  In the other places I either trusted the data to be ints (if obvious) or defensively called `parseInt` more locally (i.e. inside the loop).

showell commented:
@andersk Regarding the case of whether we fail hard for an undefined key in `FoldDict`, you are advocating for the principle of failing fast and failing clearly, which is obviously a good principle to follow in general.

I'll make the case for why being a little bit resilient here to bad data is a better approach here, and we'll see if @timabbott agrees.  Yesterday, we introduced a little bit of code to czo that relied on the invariant that every user has a non-falsy full name.  This is a variant we actually enforce, but apparently exactly one of our 13,000+ users had a full_name matching the empty string.  It was some user from 2017 who's probably never been back to the site.

That one piece of bad data took down the whole search feature.

The analogy here would be something one rogue message having a null topic for some reason, and it takes down an entire message view.  Instead of just the one piece of data failing a `get` call, and we see the blueslip error in gmail.

showell commented on static/js/people.js:
```diff
@@ -355,6 +363,7 @@ function sorted_other_user_ids(user_ids) {
         user_ids = [my_user_id];
     }
 
+    user_ids = map_int(user_ids);
```
fixed

showell commented:
Tim, everything up to but not including the first `IntDict` commit should be ready to merge.  Whatever we decide on `FoldDict` return a `TypeError` can wait, since that's a behavior change that we'll want to do as a separate commit.

andersk commented on static/js/people.js:
```diff
@@ -134,7 +137,7 @@ exports.huddle_string = function (message) {
 
     let user_ids = _.map(message.display_recipient, function (recip) {
-        return recip.id;
+        return parseInt(recip.id, 10);
```
`.id` seems like it ought to be a `number` already? If it isn’t, we should fix that.

showell replied:
fixed, thanks

andersk commented on static/js/people.js:
```diff
@@ -382,7 +387,7 @@ exports.all_user_ids_in_pm = function (message) {
 
     let user_ids = _.map(message.display_recipient, function (elem) {
-        return elem.user_id || elem.id;
+        return parseInt(elem.user_id || elem.id, 10);
```
Are `.user_id` and `.id` not already numbers?

showell commented:
One of the things that came up a couple times in this review is that we have lots of code like this:

~~~
person.user_id || person.id
~~~

This code comes from the fact that some functions could potentially operate on "pseudo" person objects like the message's `display_recipient` field.  In `person`, we call user_ids "user_id", whereas in `display_recipient` we call user_ids "id".  And then there are occasionally non-person objects like the pseudo object we create for an "all" mention in typeahead.

I didn't address all of these, but there are some functions that are clearly just dealing with `message.display_recipient` data, so I fixed those.

andersk commented on static/js/buddy_list.js:
```diff
@@ -29,7 +29,7 @@ function buddy_list_conf() {
     conf.get_key_from_li = function (opts) {
         const user_id = opts.li.expectOne().attr('data-user-id');
-        return user_id;
+        return parseInt(user_id, 10);
```
Just to add this as a diff comment so it doesn’t get forgotten: the `parseInt(elem.attr("data-user-id"), 10)` pattern should be replaced with `elem.data("user-id")` everywhere. (There are many instances of this and I won’t spam them all with comments.)

showell replied:
In buddy_list it's nice to use `data-user-id` explicitly, since it's a bit easier to grep for and we also use it to build up selectors like `conf.item_sel + "[data-user-id='" + user_id + "']";`.  It seems like it's more explicit to do `parseInt` ourselves, since we're not expecting arbitrary JSON here.

showell replied:
done as separate commit

timabbott commented:
OK, I merged the block of commits from this leading up to the introduction of `IntDict`, since those all looked good.

timabbott commented:
What's the thinking on the `IntDict` approach?  I'm OK with having an explicit class for a dictionary that requires all keys to be integers, since the extra validation that would offer seems helpful.

For the `.data()` commits, I seem to recall that a couple years ago we replaced all use of `.data(user-id)` with e.g. `.attr("data-user-id")` in part because anything set using `.data('user-id', foo)` will not be visible to `.attr("data-user-id")` while the opposite isn't true, which is very confusing.  (https://stackoverflow.com/questions/13275812/how-to-get-set-and-select-elements-with-data-attributes for details on how this jQuery feature works), so we aimed to ban use of the `.data()` setter pattern from the codebase.  

I'm OK with using that for getters to get the benefits of the type being right, but I'd like to be careful here since I know we've spent many hours in the past discovering and debugging data synchronization issues that turned out to be because using a `.data()` setter to override a `data-topic-name` after a message is edited doesn't do what one thinks.  (I think this was a 5x bigger problem before we migrated a lot of elements to use user/stream IDs rather than names in widgets, so the context has changed) .

andersk commented:
The issue is that native `Map`, rather than stringifying all keys like `Dict`, considers `string` and `number` keys to be distinct. So if we’re replacing `Dict` with `Map`, we can no longer get away with as much of the `string | number` confusion we used to have. A separate validating `IntDict` class may be necessary at this stage to root out these bugs. But hopefully we’ll be able to remove it in favor of `Map` when that rooting is done.

As for `elem.attr("data-foo")` vs. `elem.data("foo")`, I was trying to be clear that we should do all of them consistently, but Steve keeps changing just the instance I specifically commented on… It seems to me that `elem.data("foo")` is strictly cleaner than writing `parseInt(elem.attr("data-foo"), 10)` everywhere. If we’re consistent about it, we don’t have synchronization bugs, and we also don’t have bugs where we forget to parse to `number`.

showell commented:
I agree with the points Anders is making on IntDict, and I'll add that the new class is faster (there are commented-out benchmarks that you can run that show a 2x speedup) and presumably take less memory (I didn't measure that, but it seems likely).  And, to repeat a point that Anders makes, it gives us a migration path toward using pure Maps in some places.

For the `data` issue I did indeed miss a few places during the sweep.  But from a consistency standpoint, we use `attr` in lots of places.  Even if some of our original reasons for replacing `data` with `attr` no longer apply, it would be beyond the scope of this PR to fix all those, and I don't really see any clear win there.  I obviously understand the code's more compact, but I like the consistency of having the more explicit `data-user-id` attribute in all places--templates, selectors, and lookups.  Also, doing the type conversion yourself, instead of jQuery, can be more typesafe in places.

So I'm gonna go back to `parseInt`.

showell commented:
@andersk If you can convince Tim that his concerns about `data` are unfounded at this point, it's probably more efficient for us to just sweep all of these at once after this PR is merged.  There are 25 cases of `data-user-id`, and 19 cases of `data-stream-id` in `static/js`, and there are probably some unit test changes that would need to happen.

showell commented:
benchmarks (all microseconds):
* old Dict - 2.2 per set/get
* IntDict w/string conversions - 1.2 per set/get
* IntDict - 1 per set/get

So, even with the Map, using int keys gives another 20% speedup, which gets magnified a lot, since we do lots of set/get in the app.

Moving to a pure Map gives 0.8 microseconds, so we're still paying a 20% overhead for IntDict's wrappers.

timabbott commented:
>  If we’re consistent about it, we don’t have synchronization bugs, and we also don’t have bugs where we forget to parse to number.

Yeah, though one thing to be mindful of is that we do put `data-foo` elements in our HTML templates, so it's not just the JavaScript that needs to be consistent for a given widget.

timabbott commented:
I'm happy with the strategy here and didn't see any bugs, so I've merged this.  We should be sure to do a checkup on blueslip warnings from chat.zulip.org in a few days to see if this new warning appears

I'm guessing that there's some associated conversion we missed that will result in a bug, but this is as good a time as any to potentially introduce regressions in service of making the codebase more robust long-term.  

Thanks @showell and @andersk for all the work on this!
